lib/methods/ifwapichecktask.cpp

```diff
@@ -0,0 +1,400 @@
+			"Got bad perfdata from IfW API on host '" + psHost + "' port '" + psPort + "': " + ex.what(), start, end
+		);
+		return;
+	}
+
+	ReportIfwCheckResult(checkable, command, cr, result->Get("checkresult"), start, end, exitcode, perfdata);
+}
+
+void IfwApiCheckTask::ScriptFunc(const Checkable::Ptr& checkable, const CheckResult::Ptr& cr,
+	const Dictionary::Ptr& resolvedMacros, bool useResolvedMacros)
+{
+	using namespace boost::asio;
+	using namespace boost::beast;
+	using namespace boost::beast::http;
+
+	REQUIRE_NOT_NULL(checkable);
+	REQUIRE_NOT_NULL(cr);
+
+	CheckCommand::Ptr command = CheckCommand::ExecuteOverride ? CheckCommand::ExecuteOverride : checkable->GetCheckCommand();
+	auto lcr (checkable->GetLastCheckResult());
+
+	Host::Ptr host;
+	Service::Ptr service;
+	tie(host, service) = GetHostService(checkable);
+
+	MacroProcessor::ResolverList resolvers;
+
+	if (MacroResolver::OverrideMacros)
+		resolvers.emplace_back("override", MacroResolver::OverrideMacros);
+
+	if (service)
+		resolvers.emplace_back("service", service);
+	resolvers.emplace_back("host", host);
+	resolvers.emplace_back("command", command);
+
+	auto resolveMacros ([&resolvers, &lcr, &resolvedMacros, useResolvedMacros](const char* macros) -> Value {
+		return MacroProcessor::ResolveMacros(
+			macros, resolvers, lcr, nullptr, MacroProcessor::EscapeCallback(), resolvedMacros, useResolvedMacros
+		);
+	});
+
+	String psCommand = resolveMacros("$ifw_api_command$");
+	Dictionary::Ptr arguments = resolveMacros("$ifw_api_arguments$");
+	String psHost = resolveMacros("$ifw_api_host$");
+	String psPort = resolveMacros("$ifw_api_port$");
+	String expectedSan = resolveMacros("$ifw_api_expected_san$");
+	String cert = resolveMacros("$ifw_api_cert$");
+	String key = resolveMacros("$ifw_api_key$");
+	String ca = resolveMacros("$ifw_api_ca$");
+	String crl = resolveMacros("$ifw_api_crl$");
+	String username = resolveMacros("$ifw_api_username$");
+	String password = resolveMacros("$ifw_api_password$");
+
+	Dictionary::Ptr params = new Dictionary();
+
+	if (arguments) {
+		ObjectLock oLock (arguments);
+		Array::Ptr emptyCmd = new Array();
+
+		for (auto& kv : arguments) {
+			Dictionary::Ptr argSpec;
+
+			if (kv.second.IsObjectType<Dictionary>()) {
+				argSpec = Dictionary::Ptr(kv.second)->ShallowClone();
+			} else {
+				argSpec = new Dictionary({{ "value", kv.second }});
+			}
+
+			// See default branch of below switch
+			argSpec->Set("repeat_key", false);
+
+			/* MacroProcessor::ResolveArguments() converts
+			 *
+			 * [ "check_example" ]
+			 * and
+			 * {
+			 * 	 "-f" = { set_if = "$example_flag$" }
+			 * 	 "-a" = "$example_arg$"
+			 * }
+			 *
+			 * to
+			 *
+			 * [ "check_example", "-f", "-a", "X" ]
+			 *
+			 * but we need the args one-by-one like [ "-f" ] or [ "-a", "X" ].
+			 */
+			Array::Ptr arg = MacroProcessor::ResolveArguments(
+				emptyCmd, new Dictionary({{kv.first, argSpec}}), resolvers, lcr, resolvedMacros, useResolvedMacros
+			);
```

Comment:
Something is going wrong for arguments that are generated by a function. `Invoke-IcingaCheckUsers` with `vars.IcingaCheckUsers_Array_Username = ["Administrator", "Admin"]` and `vars.IcingaCheckUsers_Int32_Verbosity = 3` results in the following output (some irrelevant green dots are missing as they can't be copied from Web):
```
 Users: 1 Ok (All must be )
\_  Logged On Users "Object of type 'Function'": 0
```

Reply:
Let me guess: works fine w/o ifw-api?

Reply:
I haven't tested that.

Reply:
Which is tbh. a bit disappointing, had to re-spin-up the whole https://github.com/Icinga/icinga2/pull/9062#issuecomment-1533143736 to verify my(?) fault. (For testing the curl thing this wasn’t needed.)

But yes, it's likely my fault 😅

![Bildschirm­foto 2023-06-05 um 16 24 05](https://github.com/Icinga/icinga2/assets/6053266/78af5fa2-be4a-4c52-ac60-c42b3705f9e7)
![Bildschirm­foto 2023-06-05 um 16 27 23](https://github.com/Icinga/icinga2/assets/6053266/3e97ab85-4ab6-4e93-bbac-5c254c3ad143)

Maybe those are the _array-preprocessing in DSL functions_... (TBD)

Reply:
What are these screenshots showing? But the error in the second one looks like it may be this PR but with https://github.com/Icinga/icinga-powershell-framework/pull/629 missing? At least it looks like if the `-C` parameter was passed in the JSON and then misinterpreted as something else.

> Which is tbh. a bit disappointing

What? There shouldn't be anything surprising in being asked to test something with one's own PR. Also, there are 56 arguments that work like this, I think it's unlikely that would have gone unnoticed if it had never worked.

Reply:
> But the error in the second one looks like it may be this PR but with [Icinga/icinga-powershell-framework#629](https://github.com/Icinga/icinga-powershell-framework/pull/629) missing? At least it looks like if the `-C` parameter was passed in the JSON and then misinterpreted as something else.

Ah, yes. I "forgot" it to prove it's necessary – which I've done now 😅

Anyway: even w/ that fixed I get the same output as you.

> Also, there are 56 arguments that work like this, I think it's unlikely that would have gone unnoticed if it had never worked.

Admittedly you're right, didn’t think about that.

Reply:
## TODO

* [ ] Icinga/icinga-powershell-framework#633

Reply:
Please include relevant information like dependencies in the PR description. Something that's 70% down the PR history is way too easy to miss.

Reply:
Done.